components/authz-service/storage/postgres/migration/sql/49_update_rule_triggers.up.sql

```diff
@@ -0,0 +1,50 @@
+
+BEGIN;
+
+CREATE OR REPLACE FUNCTION update_rule(in_rule_id text, in_project_id text, in_name text, in_type text, in_deleted boolean, projects_filter TEXT[])
+RETURNS TEXT
+LANGUAGE plpgsql
+AS $$
+DECLARE
+  rule_db_id INTEGER;
+BEGIN
+    IF (NOT EXISTS (SELECT id FROM iam_project_rules ipr WHERE id=in_rule_id AND projects_match_for_rule(ipr.project_id, projects_filter))) AND
+        (NOT EXISTS (SELECT id FROM iam_staged_project_rules ipr WHERE id=in_rule_id AND projects_match_for_rule(ipr.project_id, projects_filter))) THEN
+            RAISE EXCEPTION 'not found %', in_rule_id USING
+            ERRCODE='case_not_found'; -- do we need this more specific? can we just return 20000?
+    ELSIF (NOT EXISTS (SELECT id, project_id FROM iam_project_rules ipr WHERE id=in_rule_id AND project_id=in_project_id)) AND
+        (NOT EXISTS (SELECT id, project_id FROM iam_staged_project_rules ipr WHERE id=in_rule_id AND project_id=in_project_id)) THEN
+            RAISE EXCEPTION 'incoming project does not match rule project %', in_rule_id USING -- do we need this? maybe we can just return null or something?
+            ERRCODE='PRJTR';
+    ELSE
+        INSERT INTO iam_staged_project_rules as ispr (id, project_id, name, type, deleted)
+        VALUES (in_rule_id, in_project_id, in_name, in_type, false)
+        ON CONFLICT (id) -- applied rule has already been updated, so update the staged version of it
+            DO UPDATE
+                    SET
+                    id = in_rule_id,
+                    project_id = in_project_id,
+                    name          = in_name,
+                    type   = in_type,
+                    deleted    = in_deleted
+        WHERE ispr.id = in_rule_id AND projects_match_for_rule(ispr.project_id, projects_filter)
+        RETURNING ispr.db_id INTO rule_db_id;
+    RETURN rule_db_id;
+    END IF;
+    RETURN rule_db_id;
+END;
+$$;
+
+CREATE OR REPLACE FUNCTION fn_deleted_rule_check() RETURNS TRIGGER AS $$
+BEGIN 
+    RAISE EXCEPTION 'rule with id % has been deleted', NEW.id USING
+        ERRCODE='RDLTD'; 
+END$$ LANGUAGE plpgsql;
+
+-- cannot update a rule that is staged for deletion 
```

Comment:
🤔 What if we want to update a rule to simply be undeleted? I think this is a vital use case.

Reply:
I think at this point we are not offering that option. The UI will not display rules marked for deletion, for example. 

If we get to a place where we show drafts, we can remove this.